### PR TITLE
Readable Error if torch.save(opt_model)

### PR DESCRIPTION
**Pull request by msaroufim**
Since https://github.com/pytorch/pytorch/pull/101651 is still contentious, this at least helps users not see a can't pickle convert frame error and gives an actual workaround

cc @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @aakhundov

### CODE REVIEW

File: test/dynamo/test_modules.py [hunk 1]
```diff
@@ -1818,6 +1823,26 @@ def forward(self, x):
         self.assertEqual(eager_res, optim_res)
         self.assertEqual(cnt.frame_count, 1)
 
+    def test_torch_save_error(self):
+        mod = BasicModule()
+        opt_model = torch.compile(mod)
+
+        with tempfile.TemporaryDirectory() as temp_dir:
+            file_path = os.path.join(temp_dir, "foo.pt")
+            with pytest.raises(TypeError) as excinfo:
```
Comment (albanD) [inline test/dynamo/test_modules.py:1832]: No need for pytest here. You can use `self.assertRaisesRegex()` and provide a regex for what you want to match in the error message.

File: torch/serialization.py [hunk 2]
```diff
@@ -499,7 +499,16 @@ def save(
     torch._C._log_api_usage_once("torch.save")
     _check_dill_version(pickle_module)
     _check_save_filelike(f)
-
+    from torch._dynamo import OptimizedModule
+    if isinstance(obj, OptimizedModule):
```
Comment (albanD) [inline torch/serialization.py:503]: Please move this code to OptimizedModule `__getstate__` method!

Comment (msaroufim) [reply torch/serialization.py:503]: That was my initial intent but then I can't raise an error or else that'll break the minifier and a warning will be quite noisy again for the minifier

Comment (albanD) [reply torch/serialization.py:503]: Why is the minifier involved here? Doesn't the `__call__` the only method that has a custom behavior?

**Review by jansel — CHANGES_REQUESTED** [review]
What error do you get before this PR?  I'd actually expect save(opt_model) to work already given `__setstate__`/`__getstate__` handlers.

**Pull request closed without merging**